### ISSUE TITLE
How do we send big MXF files to Glacier?

### ISSUE THREAD

**Issue by alexwlchan** [wellcomecollection/platform#4403]
The new A/V workflows are going to ingest some extremely large MXF video files (tens or hundreds of gigabytes each). They also include MP4 videos, which are much smaller and what we'll be serving most of the time.

We want to keep multiple copies of the files, but we don't want to keep the warm replica as Standard&#8209;IA – that would be expensive, and access to the MXF masters will be rare. We'd like to keep these objects in Glacier instead.

**Q1: Is this an accurate summary of the problem?**

**Comment by alexwlchan** [comment]
Here’s my best idea so far:

* Create a lifecycle policy on the `wellcomecollection-storage` bucket that moves any object with the tag `WARM_REPLICA=glacier` to the Glacier storage class after 30 days.

* Create a Lambda or ECS service that listens to the stream of new bags from the bag register. This inspects the bag manifest, and if it sees a bag that:

   - is in the digitised space
   - only has an MP4, MXF, poster image and METS file
   - has an MXF which is >5GB in size (say)

  it applies the `WARM_REPLICA=glacier` tag to the MXF object only.

Pros:

* This is Wellcome-specific logic. It shouldn't live inside the core storage service – and in this design, it doesn't!
* We can scope the service permissions very tightly; we only give it [PutObjectTagging](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketTagging.html). The bag replicator remains the only service that can write new objects to our buckets.
* We're letting Amazon do the lifecycling, rather than doing it ourselves. There are silly mistakes you can make when trying to change the storage class (e.g. accidentally replacing the object with an empty file).

Cons:

* There's no way for somebody who gets a storage manifest to know this object has been lifecycled to Glacier.

**Q2: What do people think of this approach?**

**Comment by alexwlchan** [comment]
An additional spanner in the works: if we ever upload a new version of that bag, we might refer back to the old MXF and need to use it for verification. https://wellcome.slack.com/archives/CN56BRQ5B/p1587383549362700

**Comment by kenoir** [comment]
> There's no way for somebody who gets a storage manifest to know this object has been lifecycled to Glacier.

These feels like a problem to me.

I am assuming it is necessary to store different objects in the same bag in differing storage classes as our workflow service will require immediate access to metadata in the same bag as these large files.

If this is true, should we represent those differing storage classes in the API?

 Pinging @jtweed to discuss from a product perspective.

**Comment by kenoir** [comment]
**Another question**: Should we allow the client to specify which storage tier is applied to which files? 

Could we allow the client to provide a list of file extensions or a regex to use to assign a storage tier?

**Comment by kenoir** [comment]
@alexwlchan regarding the approach you describe - I think it depends a lot on whether we choose to deny the client knowledge of what storage tier files are in. If we can do that it makes sense to separate the "Wellcome specific" knowledge as you describe.

TLDR: We need some more information from a Product perspective.

**Comment by jtweed** [comment]
So for me I think it's important that there is no implicit cleverness in the storage service. It's always been a key design principle that it just does what it's told to do.

That it turn suggests to me that the only way to support this is for services that are ingesting into the storage service to somehow specify storage tiering (whether we expose it like that or not).

The cleanest way to do this that I've ever managed to think of is to store archival only files in separate bags, in a space set up to write to Glacier rather than IA. In other words, let the workflow system explicitly decide where something goes.

I think we should talk this through on a call before going any further.

**Comment by kenoir** [comment]
@jtweed I remember our discussion about this now. Thanks for clarifying.

> The cleanest way to do this that I've ever managed to think of is to store archival only files in separate bags, in a space set up to write to Glacier rather than IA. In other words, let the workflow system explicitly decide where something goes.

This is pleasant as it is also simpler to implement.

Talking to @alexwlchan, I think there is a question about **whether the storage tier for a bag is given explicitly by the client as part of the API request** (with some default), or if something like the "space" a bag is stored in implicitly defines which storage tier is to be used.

I think again this is a question for @jtweed.

As the "whole bag/one storage tier" contains few unknowns I think this is less pressing and we can work on this in parallel to other things when we have an answer about whether to update the API.

A quick note that in this simpler scenario we can still run into issues where a "glacier" bag is updated with files added as we will be unable to verify glacier objects as part of the update. We think it's unlikely that we will immediately run into that issue.

**Comment by alexwlchan** [comment]
Summary of discussion in #4465

**Comment by alexwlchan** [comment]
Okay, here's a summary of recent discussions:

### MXF and MP4 files will be stored in the same bag

* This is easier for Intranda
* It makes more logical sense – these files belong together, so they should be stored alongside each other in S3

### A Lambda listens to the event stream of Puts to the S3 bucket, and applies a tag to large objects

* The tag will be `StorageClass=Glacier`
* This Lambda will only have the PutObjectTagging permission
* The Lambda lives separately from the main storage service, and we keep the Wellcome-specific logic out of the core
* It listens to the event stream rather than the stream of registered bags, so it can apply tags even if the bag doesn't get successfully registered

### A lifecycle rule moves all objects with the tag to Glacier

* Blocked on https://github.com/wellcomecollection/platform/issues/4464

### The verifier looks up older versions of files from their bag manifests

* If a V2 bag has a fetch.txt that refers to a file in a V1 bag, the verifier gets the V1 bag manifest to get the SHA1 checksum rather than reading the bytes directly
* This is faster and cheaper (less GetObject calls for old objects)
* It ensures that the checksum in the V2 manifest indeed matches the bytes of the V1 file
* It opens the door for a refactoring of the bag-related services (verifier, API, register) in a similar way to the ingests service – one service is responsible for talking to the bags database, and then accepted read/write requests for bags over HTTP

![bags](https://user-images.githubusercontent.com/301220/81176209-be623a00-8f9c-11ea-9d60-f5198c9ded05.jpg)

**Comment by alexwlchan** [comment]
See tickets above for how we're actually going to do this.
